### PR TITLE
LC 2249 completed learning extract

### PR DESCRIPTION
**Pull request by jaylco**
- refactored config
- Refactored database access
- Improved efficiency of duplicate token job (now deletes multiple tokens at once)
- Implemented course completions report job
    - Job finds the last successful report date from the Azure table
    - Creates an extract from the last success timestamp and todays date
    - Zips / password protects the extract
    - Sends extract and password separately via gov.uk notify

### CODE REVIEW

File: src/db/identity/database.ts [hunk 1]
```diff
@@ -1,72 +1,62 @@
...
+const COUNT_DUPLICATE_TOKENS = `SELECT authentication_id, client_id
 FROM token
 WHERE status COLLATE utf8_unicode_ci = 0
 GROUP BY authentication_id
-HAVING COUNT(*) > 1;`)[0]
-  return rows ?? []
+HAVING COUNT(*) > 1;`
+
+const DELETE_INVALID_TOKEN_SQL = `delete from token where status = 1 limit ${deleteBatchSize};`
+const DELETE_VALID_USER_TOKEN_SQL = `delete from token where status = 0 and user_name is not null limit ${deleteBatchSize};`
```
Comment (SavvasCO) [inline src/db/identity/database.ts:20]: Worth updating these as prepared statements?

File: src/service/dateService.ts
Comment (SavvasCO) [inline src/service/dateService.ts:1]: Do we need this file?

**Review by SavvasCO — COMMENTED** [review]
A few comments

File: src/db/identity/database.ts [hunk 2]
```diff
@@ -16,47 +16,47 @@ WHERE status COLLATE utf8_unicode_ci = 0
 GROUP BY authentication_id
 HAVING COUNT(*) > 1;`
 
-const DELETE_INVALID_TOKEN_SQL = `delete from token where status = 1 limit ${deleteBatchSize};`
-const DELETE_VALID_USER_TOKEN_SQL = `delete from token where status = 0 and user_name is not null limit ${deleteBatchSize};`
+const DELETE_INVALID_TOKEN_SQL = 'delete from token where status = 1 limit ?;'
+const DELETE_VALID_USER_TOKEN_SQL = 'delete from token where status = 0 and user_name is not null limit ?;'
```
Comment (SavvasCO) [inline src/db/identity/database.ts:20]: Looks fantastic! Maybe include the status field in your variables as well?

```diff
- where status = 1
+ where status = ?
```

Comment (jaylco) [reply src/db/identity/database.ts:20]: Don't think that's as necessary as we're not passing that in explicitly anywhere? It's already safe when hardcoded

**Pull request merged**